presto-main/src/test/java/com/facebook/presto/server/remotetask/TestHttpRemoteTask.java

```diff
@@ -0,0 +1,237 @@
+        taskManagerConfig.setInfoUpdateInterval(new Duration(idleTimeout.roundTo(MILLISECONDS) / 10, MILLISECONDS));
+
+        AtomicLong lastActivity = new AtomicLong();
+        HttpProcessor httpProcessor = new HttpProcessor(taskStatusCodec, taskInfoCodec, lastActivity);
+        TestingHttpClient testingHttpClient = new TestingHttpClient(httpProcessor);
+
+        HttpRemoteTaskFactory httpRemoteTaskFactory = new HttpRemoteTaskFactory(
+                new QueryManagerConfig(),
+                taskManagerConfig,
+                testingHttpClient,
+                new TestSqlTaskManager.MockLocationFactory(),
+                taskStatusCodec,
+                taskInfoCodec,
+                JsonCodec.jsonCodec(TaskUpdateRequest.class),
+                new RemoteTaskStats());
+        RemoteTask remoteTask = httpRemoteTaskFactory.createRemoteTask(
+                TEST_SESSION,
+                new TaskId("test", 1, 2),
+                new PrestoNode("node-id", URI.create("http://192.0.1.2"), new NodeVersion("version"), false),
+                TaskTestUtils.PLAN_FRAGMENT,
+                ImmutableMultimap.of(),
+                OutputBuffers.createInitialEmptyOutputBuffers(OutputBuffers.BufferType.BROADCAST),
+                new NodeTaskMap.PartitionedSplitCountTracker(i -> {
+                }),
+                true);
+
+        httpProcessor.setInitialTaskInfo(remoteTask.getTaskInfo());
+        remoteTask.start();
+        asyncRun(
+                idleTimeout,
+                failTimeout,
+                lastActivity,
+                () -> testComplete.complete(null),
+                (message, cause) -> {
+                    httpRemoteTaskFactory.stop();
+                    testComplete.completeExceptionally(new AssertionError(message, cause));
+                });
+
+        testComplete.get();
+
+        assertTrue(remoteTask.getTaskStatus().getState().isDone(), format("TaskStatus is not in a done state: %s", remoteTask.getTaskStatus()));
+        assertTrue(remoteTask.getTaskInfo().getTaskStatus().getState().isDone(), format("TaskInfo is not in a done state: %s", remoteTask.getTaskInfo()));
+    }
+
+    private static void asyncRun(Duration idleTimeout, Duration failTimeout, AtomicLong lastActivity, Runnable runAfterIdle, BiConsumer<String, Throwable> runAfterFail)
+    {
+        new Thread(new Runnable() {
+            @Override
+            public void run()
+            {
+                long startTime = System.nanoTime();
+
+                try {
+                    while (true) {
+                        if (lastActivity.get() == 0) {
+                            Thread.sleep(idleTimeout.roundTo(MILLISECONDS));
+                            continue;
+                        }
+                        long millisToIdleTarget = idleTimeout.roundTo(MILLISECONDS) - nanosSince(lastActivity.get()).roundTo(MILLISECONDS);
```

Comment:
This section is confusing.  You have a duration and are subtracting an instant, which results new negative instant.  I would assume you would do something like:
```java
long idleTimeoutTime = lastActivity.get() + idleTimeoutMillis;
long failTime = startTime + failTimoutMillis;
if (failTimeout < idleTimeoutTime) {
   fail....
}
```
It could be I'm just really confused.